server/helper_test.go

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) 2023-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"slices"
+	"testing"
+	"time"
+
+	goPlugin "github.com/hashicorp/go-plugin"
+	"github.com/mattermost/mattermost-plugin-msteams-devsecops/server/msteams"
+	"github.com/mattermost/mattermost-plugin-msteams-devsecops/server/msteams/mocks"
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/plugin"
+	pluginapi "github.com/mattermost/mattermost/server/public/pluginapi"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/oauth2"
+)
+
+type testHelper struct {
+	p                        *Plugin
+	appClientMock            *mocks.Client
+	clientMock               *mocks.Client
+	websocketClients         map[string]*model.WebSocketClient
+	websocketEventsWhitelist map[string][]model.WebsocketEventType
+}
+
+func setupTestHelper(t *testing.T) *testHelper {
+	t.Helper()
+
+	p := &Plugin{
+		// These mocks are replaced later, but serve the plugin during early initialization
+		msteamsAppClient: &mocks.Client{},
+		clientBuilderWithToken: func(redirectURL, tenantID, clientId, clientSecret string, token *oauth2.Token, apiClient *pluginapi.LogService) msteams.Client {
+			return &mocks.Client{}
+		},
+	}
+	th := &testHelper{
+		p: p,
+	}
+
+	// Set up JWK to validate JWTs from Microsoft Teams.
+	p.cancelKeyFuncLock.Lock()
+	if p.cancelKeyFunc == nil {
+		p.tabAppJWTKeyFunc, p.cancelKeyFunc = setupJWKSet()
+	}
+	p.cancelKeyFuncLock.Unlock()
+
+	// ctx, and specifically cancel, gives us control over the plugin lifecycle
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// reattachConfigCh is the means by which we get the Unix socket information to relay back
+	// to the server and finish the reattachment.
+	reattachConfigCh := make(chan *goPlugin.ReattachConfig)
+
+	// closeCh tells us when the plugin exits and allows for cleanup.``
+	closeCh := make(chan struct{})
+
+	// plugin.ClientMain with options allows for reattachment.
+	go plugin.ClientMain(
```

Comment:
Probably needs `th.appClientMock.On("GetApp", p.configuration.AppClientID).Return(&clientmodels.App{}, nil).Once()` around to mock the `checkCredentials` call when the plugin activates.